File: binary_search.py
import time
import psutil
import os

def binary_search(arr, target):
    left, right = 0, len(arr) - 1
    while left <= right:
        mid = left + (right - left) // 2
        if arr[mid] == target:
            return mid
        elif arr[mid] < target:
            left = mid + 1
        else:
            right = mid - 1
    return -1

# Funkcia na meranie pamäte
def get_memory_usage():
    process = psutil.Process(os.getpid())
    return process.memory_info().rss / 1024**2  # Prevod na MB

# Nastavenie parametrov testu
arr = [2*i for i in range(1000000)]
target = 100000

# Meranie CPU a RAM pred testom
cpu_before = psutil.cpu_percent(interval=None)
mem_before = get_memory_usage()

start = time.time()
result = binary_search(arr, target)
end = time.time()

# Meranie CPU a RAM po teste
cpu_after = psutil.cpu_percent(interval=None)
mem_after = get_memory_usage()

# Výpis výsledkov
print(f"Index hodnoty {target}: {result}")
print(f"Čas vykonania: {end - start:.6f} s")
print(f"CPU pred: {cpu_before}%, CPU po: {cpu_after}%")
print(f"RAM pred: {mem_before:.2f} MB, RAM po: {mem_after:.2f} MB")

File: binary_search.cpp
#include <iostream>
#include <vector>
#include <chrono>

using namespace std;
using namespace chrono;

int binary_search(const vector<int>& arr, int target) {
    int left = 0, right = arr.size() - 1;
    while (left <= right) {
        int mid = left + (right - left) / 2;
        if (arr[mid] == target) {
            return mid;
        } else if (arr[mid] < target) {
            left = mid + 1;
        } else {
            right = mid - 1;
        }
    }
    return -1;
}

int main() {
    vector<int> arr(1000000);
    for (int i = 0; i < 1000000; ++i) {  // Fill the array with values from 0 to 1,999,998 (even numbers)
        arr[i] = 2 * i;
    }
    
    int target = 23000;  // The value we are looking for

    auto start = high_resolution_clock::now();
    int result = binary_search(arr, target);
    auto end = high_resolution_clock::now();

    cout << "Index hodnoty " << target << ": " << result << endl;
    cout << "Execution time: " << duration<double>(end - start).count() << "s" << endl;

    return 0;
}
